sdk/src/beta9/runner/function.py

```diff
@@ -161,10 +169,13 @@ def main(channel: Channel):
             )
 
             # Invoke the function and handle its result
+            result = InvokeResult()
             try:
                 result = invoke_function(function_stub, context, task_id)
-            except BaseException as e:
-                handle_task_failure(e, gateway_stub, task_id, container_id, container_hostname)
+                if result.exception:
+                    raise result.exception
+            except BaseException:
+                handle_task_failure(result, gateway_stub, task_id, container_id, container_hostname)
                 raise
             finally:
                 monitor_future.cancel()
```

Comment:
Does the `monitor_future.cancel()` need to happen after `handle_task_failure(result, gateway_stub, task_id, container_id, container_hostname)`?

Reply:
Here's the docstring on the cancel function.

> Returns True if the future was cancelled, False otherwise. A future cannot be cancelled if it is running or has already completed.

This is almost a noop because it always has started for me, but if for some reason the future hasn't already started, we can cancel or prevent it from ever running.

I debated on even calling this function at all. Before these changes, this was called after the code thats now in `complete_task()`. Reason being that because this is within a ThreadPoolExector() context, thread_pool.shutdown() is called. I'm wondering if it makes more sense to call thread_pool.shutdown(wait=False) explicitly after `complete_task()` since the default shutdown() uses wait=True.

Reply:
The more I've read about this the more I do think it's better that we explicitly set `.shutdown(wait=False)`. I've updated the code to do this now.

Reply:
Yeah I think it will be cancelled when the thread pool context is exited anyway. Doesn't hurt to be explicit, but if omitted, I think it could be refactored to look like this: 

```python
with ThreadPoolExecutor() as thread_pool:
    monitor_future = thread_pool.submit(
        _monitor_task,
        context=context,
        stub_id=config.stub_id,
        task_id=task_id,
        container_id=container_id,
        function_stub=function_stub,
        gateway_stub=gateway_stub,
    )

    result = invoke_function(function_stub, context, task_id)

    if result.exception:
        handle_task_failure(result, gateway_stub, task_id, container_id, container_hostname)
        raise result.exception

    # End the task and send callback
    complete_task(
        gateway_stub,
        result,
        task_id,
        container_id,
        container_hostname,
        start_time,
    )
```

Reply:
Hm. That does look cleaner. Let me test that real quick.